code/controllers/subsystem/spatial_gridmap.dm

```diff
@@ -0,0 +1,256 @@
+	var/min_y = max(CEILING((center_y - range) * (1 / SPATIAL_GRID_CELLSIZE), 1), 1)//calculating these indices only takes around 2 microseconds
+
+	//the maximum x and y cell indexes to test
+	var/max_x = min(CEILING((center_x + range) * (1 / SPATIAL_GRID_CELLSIZE), 1), grid_cells_per_axis)
+	var/max_y = min(CEILING((center_y + range) * (1 / SPATIAL_GRID_CELLSIZE), 1), grid_cells_per_axis)
+
+	var/list/grid_level = grids_by_z_level[center_turf.z]
+	switch(type)
+		if(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS)
+			for(var/list/grid_row as anything in grid_level.Copy(min_y, max_y + 1))//from testing, slightly faster than iterating numbers from min_index to max_index
+				for(var/datum/spatial_grid_cell/cell as anything in grid_row.Copy(min_x, max_x + 1))
+					contents_to_return += cell.client_contents
+
+		if(SPATIAL_GRID_CONTENTS_TYPE_HEARING)
+			for(var/list/grid_row as anything in grid_level.Copy(min_y, max_y + 1))
+				for(var/datum/spatial_grid_cell/cell as anything in grid_row.Copy(min_x, max_x + 1))
+					contents_to_return += cell.hearing_contents
+
+	if(!include_center) //this is faster for things that dont care about themselves but are (probably) in the output, helps us return without using the LOS algorithm
+		contents_to_return -= center
+
+	if(!length(contents_to_return))
+		return contents_to_return
+
+	if(ignore_visibility)
+		for(var/atom/movable/target as anything in contents_to_return)
+			var/turf/target_turf = get_turf(target)
+
+			if(get_dist(center_turf, target_turf) > range)
+				contents_to_return -= target
+
+		return contents_to_return
+
+	//now that we have the first list of things to return, filter for things with line of sight to x and y
+	for(var/atom/movable/target as anything in contents_to_return)
+		var/turf/target_turf = get_turf(target)
+		var/distance = get_dist(center_turf, target_turf)
+
+		if(distance < 2)//we're adjacent so we can see it :clueless:
+			continue
+
+		if(distance > range)
+			contents_to_return -= target
+			continue
+
+		//this turf search algorithm is the worst scaling part of this proc, scaling worse than view() for small-moderate ranges and > 50 length contents_to_return
+		//luckily its significantly faster than view for large ranges in large spaces and/or relatively few contents_to_return
+		//i can do things that would scale better, but they would be slower for low volume searches which is the vast majority of the current workload
+		//maybe in the future a high volume algorithm would be worth it
+		var/turf/inbetween_turf = center_turf
+		while(TRUE)
+			inbetween_turf = get_step(inbetween_turf, get_dir(inbetween_turf, target_turf))
+
+			if(inbetween_turf == target_turf)//we've gotten to target's turf without returning due to turf opacity, so we must be able to see target
+				break
+
+			if(inbetween_turf.opacity || inbetween_turf.opacity_sources)//this turf or something on it is opaque so we cant see through it
+				contents_to_return -= target
+				break
```

Comment:
get_step(ref, diagonal dir) works

Reply:
yeah, but it'll also put you through a wall.
So
```
W = wall
V = viewer
T = target
VW
WT
```
Would incorrectly say it's fine, right? God this problem sucks.

Reply:
this should be a mapping issue since usually mappers don't make these kind of structures

Reply:
But players can ingame so it's not simply a mapping issue.

Reply:
Enforcing that on mappers wouldn't be justified. This game is a sandbox, if we need to lay down restrictions like that for our code to function our code is bad.

Also we'd need to lint maps in a really annoying way, which I don't want to do

Reply:
Does view() prefer cardinals over diagonals like our movement code?

Reply:
view works the same way as the black "dead zones" do doesn't it?

Reply:
outside of returning items outside in

Reply:
im definitely not introducing a view() call here. i might be able to think of something but right now im not sure how to make it completely accurate to view() without being view() or being slow.

Reply:
It's this problem right?
![image](https://user-images.githubusercontent.com/58055496/136124763-02cf6c61-1e2d-4f28-98cc-5aecdd57f9d4.png)
I think I've ran into it recently. Solution looked like this, see the comments on /datum/move_loop/has_target/move_towards
https://github.com/LemonInTheDark/tgstation/blob/meteor-subsystem/code/controllers/subsystem/movement/movement_types.dm#L318

Now that I look at it it still has some flaws. We're not always guaranteed to move 1 tile at a time, there should be logic to prevent x = 1, y = 0.5 sorts of things. Could also almost for sure be done better with trig functions, more a prototype then anything else.

This might also be helpful, if a bit incomplete. There's lots of people trying to solve this same issue, one of their solutions will work.

If you figure it out replace https://github.com/LemonInTheDark/tgstation/blob/meteor-subsystem/code/__HELPERS/unsorted.dm#L431. Man's a bastard

Reply:
ok, now get_hearers_in_view() works with view() again, and get_hearers_in_LOS() has been made to interface with the spatial grid. anything that isnt important to be 1:1 with view() uses the LOS version